### DISCUSSION TITLE
wrong solution behavior in unrelialbe handshake

### DISCUSSION THREAD

**Discussion by junhyeok-yoon** [ANLAB-KAIST/KENSv3#205] [Project Q&A 2. Transport Layer (TCP)] [answered]
Hello. I am trying to implement reliable data transfer regarding handshake(not yet for data trasfer).
I am pass most of the test(open, bind, handshake, close - unreliable).
But with the seed 9344135, the solution behavior is strange(especially handshake-unreliable, TestAccept_MultipleInterface1) as below.
Not only this seed, some other seeds has these behavior(such as 7628041...) But still "pass" the test
(It keeps retransmit the FINACK)
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/114990411/2bdd7366-d6e3-4fc7-9d17-fbb6b1d287d6)
And my solution acts like the solution(which keep transmit the FINACK).
I cannot detect FINACK at packetArrived() fuction which means that the ACK should be sent from the solution.
How should I handle this problem?

**Comment by Pusnow** [comment]
It looks very strange behavior. Would you send your source code to TAs? cs341.pj@kaist.ac.kr

Until I investigate this issue, you can ignore packets after packet No. 60. It seems all connections are closed well, and you passed all test cases.

**Comment by Pusnow** [comment]
It turns out your TCP implementation keeps retransmitting the packets in line 1623.
Please check your timerCallback implementation.

**Reply by junhyeok-yoon** [reply]
FINACK을 Retransmit하고 다시 ACK을 받아야하기 때문에, 다시 타이머를 설정해줍니다. 그래서 해당 ACK이 오면 line 1401에서 그 timer를 cancel해주는데, 그 ACK이 다시 안오는 것 같습니다

**Reply by Pusnow** [reply]
I've figured out it is a bug in KENS's TCP. I'm fixing it now, and let you know when it is fixed.

**Reply by Pusnow** [reply, marked as answer]
I've fixed and uploaded new solution files. Please rebuild and test your project from scratch (remove `build/`).

If you encounter a new problem due to the changes, you can revert to old files by changing version to 3.3.3 at below.

https://github.com/ANLAB-KAIST/KENSv3/blob/5bf97d4d228d6a6bcb8a2fdc1f50abe9c075c701/CMakeLists.txt#L3

**Reply by junhyeok-yoon** [reply]
Thank you very much! By updating the version, it works well!
I tested 10000 random seeds and pass all of unreliable handshake, close.

BTW, I wonder why only fixing the version to 3.3.4 can fixed the bug.(without any git pull) How is it possible?

**Reply by Pusnow** [reply]
It was bug in the solution binary, which is distributed separately via web.

**Reply by junhyeok-yoon** [reply]
I appreciate your kind help!
